### PR TITLE
Add develop version to ufs-weather-model (major updates)

### LINKED ISSUE #377: Fix the ufs-weather-model package at spack
Fix the ufs-weather-model package at spack

### PR DESCRIPTION
**Pull request by AlexanderRichert-NOAA**
This PR adds a develop version to the ufs-weather-model recipe, which includes major updates to variants and dependencies. Tested with ATM, S2SW, and ATMAERO configurations to cover all dependencies.

Fixes https://github.com/JCSDA/spack-stack/issues/377

### CODE REVIEW

File: var/spack/repos/builtin/packages/ufs-weather-model/package.py
Comment (t-brown) [inline var/spack/repos/builtin/packages/ufs-weather-model/package.py]: Happy to be removed as a maintainer (since I'm not really doing anything!).

File: var/spack/repos/builtin/packages/ufs-weather-model/package.py
Comment (t-brown) [inline var/spack/repos/builtin/packages/ufs-weather-model/package.py:143]: Is this indented too far? Did you run it through the spack format-er (`spack style`)?

Comment (AlexanderRichert-NOAA) [reply var/spack/repos/builtin/packages/ufs-weather-model/package.py:143]: Yep, this was what `spack style --fix` gave me.

**Review by t-brown — APPROVED** [review]

File: var/spack/repos/builtin/packages/ufs-weather-model/package.py
Comment (becker33) [inline var/spack/repos/builtin/packages/ufs-weather-model/package.py]: You should keep the commit information around on the tags -- that allows Spack to create binary caches for the package and actually know whether two versions are the same.

File: var/spack/repos/builtin/packages/ufs-weather-model/package.py
Comment (becker33) [inline var/spack/repos/builtin/packages/ufs-weather-model/package.py]: This should probably be controlled by whether `avx2` is in the target, not by the variant. As currently written this could concretize for avx2 and a target that doesn't support it.

IIRC you can make queries like `if "avx2" in self.target"` to check for things like this.

File: var/spack/repos/builtin/packages/ufs-weather-model/package.py
Comment (becker33) [inline var/spack/repos/builtin/packages/ufs-weather-model/package.py]: What architectures is this for supporting? This may also be better handled in the target.

File: var/spack/repos/builtin/packages/ufs-weather-model/package.py
Comment (becker33) [inline var/spack/repos/builtin/packages/ufs-weather-model/package.py:51]: Is there a reason for these options to only be available on develop?

File: var/spack/repos/builtin/packages/ufs-weather-model/package.py
Comment (becker33) [inline var/spack/repos/builtin/packages/ufs-weather-model/package.py:44]: Should this variant be restricted to the same versions as `cccp`?

File: var/spack/repos/builtin/packages/ufs-weather-model/package.py
Comment (becker33) [inline var/spack/repos/builtin/packages/ufs-weather-model/package.py]: Are these all truly `@develop` or is this in lieu of something like `@3:` where we just don't know what the next release number is yet?

**Review by becker33 — DISMISSED** [review]

File: var/spack/repos/builtin/packages/ufs-weather-model/package.py
Comment (AlexanderRichert-NOAA) [reply var/spack/repos/builtin/packages/ufs-weather-model/package.py:51]: All of the variant with 'when="@develop"' are newer options not found in the older versions.

File: var/spack/repos/builtin/packages/ufs-weather-model/package.py
Comment (AlexanderRichert-NOAA) [reply var/spack/repos/builtin/packages/ufs-weather-model/package.py]: More or less- My expectation is that the typical use case for this package will be development off of the develop branch, which is why I haven't put in other versions.

File: var/spack/repos/builtin/packages/ufs-weather-model/package.py
Comment (AlexanderRichert-NOAA) [reply var/spack/repos/builtin/packages/ufs-weather-model/package.py]: Is there a way to create a conflict that encapsulates that? Since it's a cmake option, I think it would make sense to still allow users to disable it for debugging purposes.

File: var/spack/repos/builtin/packages/ufs-weather-model/package.py
Comment (AlexanderRichert-NOAA) [reply var/spack/repos/builtin/packages/ufs-weather-model/package.py:44]: There's no package dependency here, if that's what you mean. ccpp comes through one of the submodules.

File: var/spack/repos/builtin/packages/ufs-weather-model/package.py
Comment (AlexanderRichert-NOAA) [reply var/spack/repos/builtin/packages/ufs-weather-model/package.py]: This is a cmake option that adds '-axSSE4.2,CORE-AVX2' to cmake cflags/fflags.

File: var/spack/repos/builtin/packages/ufs-weather-model/package.py
Comment (becker33) [reply var/spack/repos/builtin/packages/ufs-weather-model/package.py:44]: No, I mean that you have a `when` restriction on the `cccp` variant, but not on this seemingly related variant.

File: var/spack/repos/builtin/packages/ufs-weather-model/package.py
Comment (becker33) [reply var/spack/repos/builtin/packages/ufs-weather-model/package.py]: You could set it up with a when-clause, but it would be a bit clunky

File: var/spack/repos/builtin/packages/ufs-weather-model/package.py
Comment (becker33) [reply var/spack/repos/builtin/packages/ufs-weather-model/package.py]: Ok yes this should be combined with the previous and use the `target` -- you can always target an earlier architecture if you don't want those features

File: var/spack/repos/builtin/packages/ufs-weather-model/package.py
Comment (becker33) [reply var/spack/repos/builtin/packages/ufs-weather-model/package.py:51]: Yes, my point is they aren't specific to the develop branch -- if you want to be forward-thinking, you can make them be `@2.1:` so that there is less work to do when future versions with these features are added.

But if you expect develop to be the only version users care about I guess it doesn't matter.

File: var/spack/repos/builtin/packages/ufs-weather-model/package.py
Comment (AlexanderRichert-NOAA) [reply var/spack/repos/builtin/packages/ufs-weather-model/package.py:44]: Ah I see. No, this tracks with what's in the cmake config for each version.

File: var/spack/repos/builtin/packages/ufs-weather-model/package.py
Comment (AlexanderRichert-NOAA) [reply var/spack/repos/builtin/packages/ufs-weather-model/package.py:51]: Yeah in practice it'll only be used for develop. I'm not even sure if anyone currently uses the numbered versions here, so I'll probably take those out at some point down the road just to clean things up.

File: var/spack/repos/builtin/packages/ufs-weather-model/package.py
Comment (AlexanderRichert-NOAA) [reply var/spack/repos/builtin/packages/ufs-weather-model/package.py]: Would this basically involve creating an entry for each CPU architecture (`when="target=x86_64_v3"`, `when="target=haswell"`, etc.)?

File: var/spack/repos/builtin/packages/ufs-weather-model/package.py
Comment (becker33) [reply var/spack/repos/builtin/packages/ufs-weather-model/package.py:47]: You can do ranges of targets, so you could say `when="target=:haswell"` or whatever the appropriate target is. But you'd have to have a check for the intel cpus, a check for the amd cpus, and checks for ppc and arm.

Much better to just trigger it by whether the arch supports it and build for the older arch if you don't want to use it.

File: var/spack/repos/builtin/packages/ufs-weather-model/package.py
Comment (becker33) [reply var/spack/repos/builtin/packages/ufs-weather-model/package.py]: You can do ranges of targets, so you could say when="target=:haswell" or whatever the appropriate target is. But you'd have to have a check for the intel cpus, a check for the amd cpus, and checks for ppc and arm.

Much better to just trigger it by whether the arch supports it and build for the older arch if you don't want to use it.

Comment (AlexanderRichert-NOAA) [reply var/spack/repos/builtin/packages/ufs-weather-model/package.py]: Since those settings get used for specific sites anyway (at those sites ufs wm will not be installed through spack), I'm setting those cmake options to OFF and removing the variants so that the target settings will be set purely through spack.

File: var/spack/repos/builtin/packages/ufs-weather-model/package.py
Comment (AlexanderRichert-NOAA) [reply var/spack/repos/builtin/packages/ufs-weather-model/package.py]: see comment in other thread about setting target purely through spack

**Comment by AlexanderRichert-NOAA** [comment]
@becker33 I took out the architecture settings since I don't think we need them for the spack recipe anyway (they're for configuring specific sites which we don't care about), so this way it'll just configure those settings through SPACK_TARGET_ARGS.

File: var/spack/repos/builtin/packages/ufs-weather-model/package.py
Comment (adamjstewart) [reply var/spack/repos/builtin/packages/ufs-weather-model/package.py:51]: I have the same concerns as @becker33 here. Spack defaults to stable checksummable releases. We still want to keep these, and add new versions in the future. Even if you only use develop, other users will likely prefer a stable release so they can reproduce the results of their research. My suggestion would be to use `@2.1:` for things that were added in 2.1, `@:2.0` for things that exist in 2.0.X but don't in 2.1.X, etc. But you're the package maintainer, so technically you get the final say in what you want to support. So if you only care about develop and no one else volunteers to support these stable releases, you are technically allowed to deprecate and remove them.

File: var/spack/repos/builtin/packages/ufs-weather-model/package.py
Comment (adamjstewart) [reply var/spack/repos/builtin/packages/ufs-weather-model/package.py:143]: Indentation is correct, it's inside a for-loop.

File: var/spack/repos/builtin/packages/ufs-weather-model/package.py
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/ufs-weather-model/package.py]: The default `type` is build/link. Is Python also needed in the `PATH` or `PYTHONPATH` at run-time?

File: var/spack/repos/builtin/packages/ufs-weather-model/package.py
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/ufs-weather-model/package.py]: Doesn't this depend on your OS? Like, a user can't choose this themselves, right? If you're on Linux and using the Intel compiler, you aren't allow to use any other choice. It might be better to auto-detect this always and remove the variant.

File: var/spack/repos/builtin/packages/ufs-weather-model/package.py
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/ufs-weather-model/package.py]: If the variant name and cmake flag name are always the same, you could do a convenient for-loop like:
```python
args = []
variants = ["debug", "moving_nest", ...]
for variant in variants:
    args.append(from_variant(variant.upper(), variant))
```
But this is completely optional, whatever is easiest for you to understand and maintain.

File: var/spack/repos/builtin/packages/ufs-weather-model/package.py
Comment (AlexanderRichert-NOAA) [reply var/spack/repos/builtin/packages/ufs-weather-model/package.py:51]: The tricky thing here is that this model doesn't really follow regular, sequential releases. In fact, there hasn't been a "regular" release since 2.0. All the other tags/releases are specific to configurations that we're not trying to support through the Spack recipe (for instance, there are tags for operational configurations that are only used internally to NOAA, and those will not be installed with Spack). Would it be okay to leave it as is, but with the goal of changing these settings to specific versions if/when new versions emerge?

**Review by adamjstewart — COMMENTED** [review]
I'm happy to take over reviewing for @becker33. My B.S. was in Earth Science and I'm now working on ML weather forecasting, so I have some interest in seeing UFS improved.

File: var/spack/repos/builtin/packages/ufs-weather-model/package.py
Comment (adamjstewart) [reply var/spack/repos/builtin/packages/ufs-weather-model/package.py:51]: I'm fine with leaving it as is for now. Are you a developer of UFS? Maybe you can encourage the other devs to make more regular releases in the future so the versions have some kind of meaning. This is really important for experiment reproducibility.

File: var/spack/repos/builtin/packages/ufs-weather-model/package.py
Comment (AlexanderRichert-NOAA) [reply var/spack/repos/builtin/packages/ufs-weather-model/package.py]: done

File: var/spack/repos/builtin/packages/ufs-weather-model/package.py
Comment (AlexanderRichert-NOAA) [reply var/spack/repos/builtin/packages/ufs-weather-model/package.py]: No, it's just needed at build time. I just set `type="build"`.

File: var/spack/repos/builtin/packages/ufs-weather-model/package.py
Comment (AlexanderRichert-NOAA) [reply var/spack/repos/builtin/packages/ufs-weather-model/package.py]: Yeah I agree this is nicer, I just made this change.

**Review by adamjstewart — APPROVED** [review]
I don't see any other red flags. As long as it builds for you and you're committed to helping fix any bugs that other users report, it should be fine. Hoping we can see more regular releases in the future!

**Pull request merged**
